controllers/helpers.go

```diff
@@ -597,3 +600,46 @@ func GetClusterIdentityFromRef(ctx context.Context, c client.Client, azureCluste
 	}
 	return nil, nil
 }
+
+func clusterIdentityFinalizer(prefix, clusterNamespace, clusterName string) string {
+	return fmt.Sprintf("%s/%s-%s", prefix, clusterNamespace, clusterName)
+}
+
+// EnsureClusterIdentity ensures that the identity ref is allowed in the namespace and sets a finalizer.
+func EnsureClusterIdentity(ctx context.Context, c client.Client, object conditions.Setter, identityRef *corev1.ObjectReference, finalizerPrefix string) error {
+	namespace := object.GetNamespace()
+	identity, err := GetClusterIdentityFromRef(ctx, c, namespace, identityRef)
+	if err != nil {
+		return err
+	}
+	if !scope.IsClusterNamespaceAllowed(ctx, c, identity.Spec.AllowedNamespaces, namespace) {
+		conditions.MarkFalse(object, infrav1.NetworkInfrastructureReadyCondition, infrav1.NamespaceNotAllowedByIdentity, clusterv1.ConditionSeverityError, "")
+		return errors.New("AzureClusterIdentity list of allowed namespaces doesn't include current cluster namespace")
+	}
+	identityHelper, err := patch.NewHelper(identity, c)
+	if err != nil {
+		return errors.Wrap(err, "failed to init patch helper")
+	}
+	// If the AzureClusterIdentity doesn't have our finalizer, add it.
+	controllerutil.AddFinalizer(identity, clusterIdentityFinalizer(finalizerPrefix, object.GetNamespace(), object.GetName()))
+	// Register the finalizer immediately to avoid orphaning Azure resources on delete.
+	return identityHelper.Patch(ctx, identity)
+}
+
+// RemoveClusterIdentityFinalizer removes the finalizer on an AzureClusterIdentity.
+func RemoveClusterIdentityFinalizer(ctx context.Context, c client.Client, object client.Object, identityRef *corev1.ObjectReference, finalizerPrefix string) error {
+	identity, err := GetClusterIdentityFromRef(ctx, c, object.GetNamespace(), identityRef)
```

Comment:
nit: slight preference for assigning `object.GetNamespace()` to a var like in `EnsureClusterIdentity` as we need it again below when we call `controllerutil.RemoveFinalizer`